server/src/main/java/io/druid/curator/discovery/CuratorDruidLeaderSelector.java

```diff
@@ -0,0 +1,204 @@
+
+  private final LifecycleLock lifecycleLock = new LifecycleLock();
+
+  private final DruidNode self;
+  private final CuratorFramework curator;
+  private final String latchPath;
+
+  private DruidLeaderSelector.Listener listener = null;
+  private final AtomicReference<LeaderLatch> leaderLatch = new AtomicReference<>();
+
+  private volatile boolean leader = false;
+  private volatile int term = 0;
+
+  public CuratorDruidLeaderSelector(CuratorFramework curator, @Self DruidNode self, String latchPath)
+  {
+    this.curator = curator;
+    this.self = self;
+    this.latchPath = latchPath;
+  }
+
+  private LeaderLatch createNewLeaderLatch()
+  {
+    final LeaderLatch newLeaderLatch = new LeaderLatch(
+        curator, latchPath, self.getHostAndPortToUse()
+    );
+
+    newLeaderLatch.addListener(
+        new LeaderLatchListener()
+        {
+          @Override
+          public void isLeader()
+          {
+            try {
+              if (leader) {
+                log.warn("I'm being asked to become leader. But I am already the leader. Ignored event.");
+                return;
+              }
+
+              leader = true;
+              term++;
+              listener.becomeLeader();
+            }
+            catch (Exception ex) {
+              log.makeAlert(ex, "listener becomeLeader() failed. Unable to become leader").emit();
+
+              // give others a chance to become leader.
+              final LeaderLatch oldLatch = createNewLeaderLatch();
+              CloseQuietly.close(oldLatch);
+              leader = false;
+              try {
+                //Small delay before starting the latch so that others waiting are chosen to become leader.
+                Thread.sleep(1000);
+
+                leaderLatch.get().start();
+              }
+              catch (Exception e) {
+                // If an exception gets thrown out here, then the node will zombie out 'cause it won't be looking for
+                // the latch anymore.  I don't believe it's actually possible for an Exception to throw out here, but
+                // Curator likes to have "throws Exception" on methods so it might happen...
+                log.makeAlert(e, "I am a zombie").emit();
+              }
+            }
+          }
+
+          @Override
+          public void notLeader()
+          {
+            try {
+              if (!leader) {
+                log.warn("I'm being asked to stop being leader. But I am not the leader. Ignored event.");
+                return;
+              }
+
+              leader = false;
+              listener.stopBeingLeader();
+            }
+            catch (Exception ex) {
+              log.makeAlert(ex, "listener.stopBeingLeader() failed. Unable to stopBeingLeader").emit();
+            }
+          }
+        },
+        Execs.singleThreaded(StringUtils.format("LeaderSelector[%s]", latchPath))
```

Comment:
This is not lifecycled at all. is it possible to have the lifecycle here controlled somehow, to make sure that the executor is not running when this jvm is not the leader?

Reply:
Gotcha, I misunderstood the scope of this

Reply:
@himanshug wait, this recurses via https://github.com/druid-io/druid/pull/4699/files#diff-2ed9a00e63ba80d914302e0f4b2a18b4R93 so it can have a lot of these, right?

Reply:
hmmm, given that this code existed before and has been working fine, earlier I though that these would become garbage and GC'd away as no reference to these executors is being maintained.
however I did some tests today to verify that and it appears that yes these objects don't get GC'd and all of them linger in the jvn, don't cause anything bad but waste some memory. I will update the code to fix this. thanks.

Reply:
@drcrallen updated code to not recreate executor every time but just once.